### PR TITLE
[DeviceMesh] Add a _MeshEnv attr to record the mapping of flatten mesh_dim_name to its mesh dim index in root mesh

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133839
* __->__ #133838

```
# supposed we have a 3d mesh
mesh_3d = init_device_mesh("cuda", (2,2,2), mesh_dim_names=("dp", "cp", "tp")
dp_cp_mesh = mesh_3d["dp", "cp"]._flatten()

"""
then we would have
flatten_name_to_root_dims[mesh_3d]: {
    "dp_cp": (0, 1)
}
"""
```

We need this information to validate the order mesh slice including flatten mesh dim.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @tianyu-l

### CODE REVIEW

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -152,12 +156,13 @@ def create_flatten_mesh(self, device_mesh: "DeviceMesh") -> "DeviceMesh":
...
                     mesh_nd,
-                    mesh_dim_names=(flatten_mesh_dim_names,),
+                    mesh_dim_names=(flatten_mesh_dim_name,),
                 )
                 if cur_rank in mesh_nd:
                     res_flattened_mesh = flattened_mesh
             self.child_to_root_mapping[res_flattened_mesh] = root_mesh  # type: ignore[possibly-undefined]
-            self.root_to_flatten_mapping.setdefault(root_mesh, {})[flatten_mesh_dim_names] = res_flattened_mesh  # type: ignore[possibly-undefined]
+            self.root_to_flatten_mapping.setdefault(root_mesh, {})[flatten_mesh_dim_name] = res_flattened_mesh  # type: ignore[possibly-undefined]
+            self.flatten_name_to_root_dims.setdefault(root_mesh, {})[flatten_mesh_dim_name] = tuple(flatten_dims_in_root)  # type: ignore[possibly-undefined]
```
Comment (fegin) [inline torch/distributed/device_mesh.py:165]: It seems that we do not check the existence of `flatten_mesh_dim_name`. If it exist, shouldn't the `create_flatten_mesh()` just returns the cached one?

File: test/distributed/test_device_mesh.py [hunk 2]
```diff
@@ -579,6 +579,13 @@ def test_flatten_mesh(self):
         dp_cp_mesh = mesh_3d["dp", "cp"]
         flattened_dp_cp_mesh = dp_cp_mesh._flatten()
         self.assertEqual(dp_cp_mesh.mesh.flatten(), flattened_dp_cp_mesh.mesh)
+        self.assertEqual(flattened_dp_cp_mesh.mesh_dim_names[0], "dp_cp")
+        root_mesh = _mesh_resources.get_root_mesh(dp_cp_mesh)
+        self.assertEqual(root_mesh, mesh_3d)
+        flatten_mesh_root_dims = _mesh_resources.flatten_name_to_root_dims[root_mesh][
+            "dp_cp"
+        ]
+        self.assertEqual(flatten_mesh_root_dims, (0, 1))
```
Comment (fegin) [inline test/distributed/test_device_mesh.py:588]: Can we also test if we `flatten()` again, we will get the same one?

Comment (wz337) [reply test/distributed/test_device_mesh.py:588]: Added the check in line 593 (also included checked for pg reuse).

File: torch/distributed/device_mesh.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/device_mesh.py:165]: > It seems that we do not check the existence of `flatten_mesh_dim_name`. If it exist, shouldn't the `create_flatten_mesh()` just returns the cached one?

Ye! Check for existence already exists in line 136 - 141.

**Review by fegin — APPROVED** [review]
LGTM

**Pull request closed without merging**